### ISSUE TITLE
Prevent display of NSFW content in Local and Federated timelines

### ISSUE THREAD

**Issue by ThisIsMissEm** [tootsuite/mastodon#6970]
Presently the "Always show media marked as sensitive" feature isn't cutting it. What people mean, is "show me sensitive media from people I follow" but what they get is EVERYTHING. This is a particular problem for switter.at, where we do have some users posting scat content, it is correctly marked as NSFW, but because a lot of users turn on the "Always show media marked as sensitive", they get subjected to it.

Idea is to not show all media marked as sensitive in local & federated timelines.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by nynhex** [comment]
I vote to +1 this.  While I have no issues with switter.at content (as I don't offend easily), there are users in the fedeverse will feel differently.

If nobody wants to pick this up, I will and start a PR.

**Comment by Aldarone** [comment]
I think this will be waayyyyyy too confusing for the users.

People who want to see all nsfw media will wonder why they are not seeing some of them. And they're gonna ask for the checkbox to really see all nsfw media.

Plus, it doesn't really make sense. Why the media of the people I follow should be seen as less nsfw than the ones of the people I don't follow?

**Comment by deutrino** [comment]
I wanted to chime in and add a bit of nuance to the OP's suggestion:

> What people mean, is "show me sensitive media from people I follow"

No, what I meant was "show me all sensitive media."

However, I think the following compromises might be good starting points for discussion: 

  * Always show sensitive media from people I follow and anything those people boost
  * Always show sensitive media from everyone in all contexts except the federated timeline
  * Always show sensitive media from everyone in every context except from certain instances (adjacent: #6962), perhaps selectable by the user and/or by the admin 
  * Revisit the community tagging idea in #318

**Comment by Cassolotl** [comment]
At the moment the [switter.at code of conduct](https://switter.at/about/more) says:

> NSFW content should be flagged with the CW button for text, and the eye button for images.
> 
> Avatars and header images should be SFW.

I've seen a toot going around saying that this rule applies to toots that are crossposts from Twitter, which needs to be said since no one has made a crossposter yet that lets you mark media as sensitive in the Twitter to Mastodon direction.

It's useful to know that if you put #nsfw in a toot it'll mark the media as sensitive, even if it's a crossposted tweet. Perhaps switter.at could add this information to their guidance on NSFW content?

At the moment the onus is on the switter.at mods to enforce their rule, which may be a very big job since there's ~10,000 new members who've joined in the past 3 days! There's also stuff to think about like, I've seen people complaining that people are getting around the NSFW rule by covering nipples with something, when actually a mostly-naked person with only their nipples covered is still very much NSFW and it should be marked sensitive. Everything is very new and the social norms haven't settled yet, you know?

I think if "always show media marked as sensitive" didn't apply to a public timeline that would be very confusing. Rearranging and rewording the setting could be helpful:

[ ] Always show media marked as sensitive everywhere
[ ] Always show media marked as sensitive in the local timeline and Home
[ ] Always show media marked as sensitive in Home

But I think really we should give switter.at a week or two to find some good community rules, spread the word about them, and find a good way to enforce them really well, before we start adding complexity to this setting.

**Comment by Coffeeadikt** [comment]
I like **Cassolotl**'s idea and would like to add that it'd be nice to have an option to auto-expand CW'd posts and not just media.
I guess I have the opposite sensitivity problem - it costs me an extra click (because I want to see everything) and those extra clicks add up very quickly in an active Local or Federated timeline. If we're fine-tuning options it'd be nice to have that as a choice too :)

**Comment by ThisIsMissEm** [comment]
"people complaining that people are getting around the NSFW rule by covering nipples with something" — this actually makes it SFW, it doesn't matter that you know what's behind the covering, it's actually still safe for work. I had a long chat with a friend/facebook moderator the other night about this, and it turns out for them, if the nipple is entirely in shadow and hard to be seen, then it's actually not needing removal. 

Moderating content is always going to be a challenge, however, I feel like we need more than just the option we have currently.

Crossposts from twitter are problematic for other reasons — we've had a LOT of spam reports about crossposts from twitter, and honestly, I kind of agree: we want there to be a community at switter, not just to be an echo of twitter. I'm beginning to think crossposting from twitter to switter is just spammy, going the other way seems mostly fine though.

I like the UI mockup that @Cassolotl did, and it's much within what I was thinking.

For Switter, we do have some members who post scat or fully nude penises, correctly marked as NSFW, however, I want to see the NSFW content from my friends that I follow, I really don't want to see some random guys dick. This is why I think we need more nuanced visibility controls.

**Comment by ThisIsMissEm** [comment]
Additionally, switter has made a modification to mark all outbound content as NSFW, regardless of what it is on switter — sure, this is not the best solution, but it's the best we can do for good federation with other instances. What we allow and what you allow may be very different things. (Ideally this would be an actual upstream change, but for now it is not)

**Comment by ThisIsMissEm** [comment]
> @Coffeeadikt 
> I guess I have the opposite sensitivity problem - it costs me an extra click (because I want to see everything) and those extra clicks add up very quickly in an active Local or Federated timeline. If we're fine-tuning options it'd be great to have that as a choice too :)

Yeah, some people may want to really see everything, but I'm not joking that we have everything. So, that's where I'm thinking more fine-grained controls would be better. Like, honestly, I can't imagine what it'd be like to see _everything_ on twitter, and quite frankly, I wouldn't want to.

**Comment by nightpool** [comment]
>Ideally this would be an actual upstream change, but for now it is not


I think there are is a discrete subsection of changes that makes a ton of sense for switter but doesn't make sense for mastodon-as-a-product. I don't know if upstreaming always needs to be the endgoal of every piece of development.

**Comment by ThisIsMissEm** [comment]
@nightpool we ideally don't want to diverge heavily from the core mastodon codebase, as that makes updating to newer versions harder. Sure, other communities might not have the same volume of adult content and moderation issues as ours, but welcome to the diversity of the internet. Ideally, lessons learned with switter should be able to be taken forth into other communities too.

e.g., maybe a community has a thing for guns and posting lots of graphically violent imagery. They too should be able to just blanket prevent their imagery from coming up unmarked from the federated timeline.

However, the original purpose of this issue is just having more fine-grained sensitive content settings, already we're seeing cases of people marking their dick pics correctly as NSFW, but almost everyone has "always show media marked as sensitive" turned on, because they want to see the media in their own and others who they follow's toots, but don't expect to see scat/dick pics.

This is only a problem mastodon has, because mastodon has the local / federated timelines. Twitter doesn't, therefore it's not an issue there and that setting makes sense on twitter. The current setting on mastodon does not make sense given what you're exposed to through the local / federated timelines.

**Comment by nightpool** [comment]
> but welcome to the diversity of the internet

Sure, but all that i'm saying is just that i wouldn't be too worried about a *little* divergence. Most instances I know of have a little divergence, and it's healthy, because it shows that the software is being adapted to fit the community, not the other way around.

no single product can fit all people.

Anyway, this isn't a comment about this issue, I'm perfectly happy to see a dropdown "Show all sensitive media by default: [                    ]" replace our current boolean setting.

**Comment by ThisIsMissEm** [comment]
Oh, on this @Cassolotl:
> I've seen a toot going around saying that this rule applies to toots that are crossposts from Twitter, which needs to be said since no one has made a crossposter yet that lets you mark media as sensitive in the Twitter to Mastodon direction.

I've proposed to the maker of the one of the crossposters implement a system that enables instance admins to opt-out / enforce certain policies (did this via mastodon, not github):
> implement a `/.well-known/crossposter` which defines config for crossposter to use (rate limits, default content warning / nsfw, prohibited accounts, etc)

**Comment by Cassolotl** [comment]
@ThisIsMissEm Awesome! @renatolond did this: *When creating status, if no sensitive status is given, use default #7057* - I'm not 100% sure but I think this would fix the issue of crossposters posting NSFW media without the media hidden even though the account default is to hide media? Correct me if I'm wrong though!

**Comment by ThisIsMissEm** [comment]
@Cassolotl the only problem now left is there are no admin tools to revoke all existing crossposters: there are multiple versions deployed, and some are always sending sensitive = false.

I've also an issue open with some code to make this change work from the client-side so that "sensitive" is only sent if the user explicitly selected one, otherwise it defaults to what the server thinks is correct. (though I'm a bit stuck on it).

**Comment by SoniEx2** [comment]
These are serious issues for federation and is it just me or does nobody seem to be working on them?

Covering up nipples with a bit of tape or in the shadow doesn't make them universally SFW. There are a wide range of jobs out there. If you work with kids, for example, it'd still be very much NSFW. Have you considered that?

**Comment by CorruptMonkey** [comment]
Anything happening with this? +1, bump, etc.

**Comment by colinstu12** [comment]
This still seems desirable?
